src/shared/ui/picker.ts

```diff
@@ -145,3 +148,213 @@ export function verifySinglePickerOutput<T extends vscode.QuickPickItem>(choices
 
     return choices[0]
 }
+
+// TODO: Cache these results? Should we have a separate store? Can we also use a store with values from the explorer tree?
+// If we move this to a cache, we should remove the awsCallLogic and instead listen/send events to/from the cache.
+// (this should also be retooled to not use the request/response objects directly)
+// TODO: Add manual pagination?
+/**
+ * Grants a `vscode.QuickPick` iteration capabilities:
+ * * Pause/unpause loading
+ * * Refresh capabilities
+ * * No item/error nodes
+ *
+ * External control is still done via `picker.promptUser` as if it were a normal QuickPick.
+ */
+export class IteratingQuickPickController<TResponse> {
+    private isDone: boolean = false
+    private isPaused: boolean = false
+    private iterator: AsyncIterator<vscode.QuickPickItem[]>
+    private activeIterationTime: Date
```

Comment:
If you can, it'd be useful to pull all re-settable fields into a separate class. That way, you prevent bugs when someone adds a new field and forgets to reset it later.

Your constructor would just do `this.state = new State()`. And your reset method would just do `this.state = new State()`

Alternatively, the reset method does `this.state.reset()`.